File: ocaml/lib/parser.ml
open Base
open Tokenizer
open Base.Option

type expr = Num of int | Binop of expr * operator * expr

let precedence (op : operator) =
  match op with
  | Add -> 1
  | Subtract -> 1
  | Multiply -> 2
  | Divide -> 2
  | Exponentiate -> 3

let prefix tokens =
  match tokens with
  | [] -> None
  | TokenNum n :: rest -> Some (Num n, rest)
  | TokenOp _ :: _ -> None

let rec pratt precLimit tokens =
  match prefix tokens with
  | None -> None
  | Some (left, remainder) -> prattLoop precLimit left remainder

and prattLoop precLimit left remainder =
  match remainder with
  | TokenOp op :: tokensAfterOp ->
      let opPrec = precedence op in
      if opPrec > precLimit then
        match pratt opPrec tokensAfterOp with
        | Some (right, tokensAfterChild) ->
            let newLeft = Binop (left, op, right) in
            prattLoop precLimit newLeft tokensAfterChild
        | None -> None
      else Some (left, remainder)
  | _ -> Some (left, remainder)

let parse tokens =
  let f tokens =
    match pratt 0 tokens with
    | Some (x, _) -> Some x
    | None -> None
  in
  tokens >>= f

File: rust/src/parser.rs
use std::cell::{Cell, RefCell};
use std::collections::VecDeque;
use crate::tokenizer::{Operation, Token};

fn precedence(op: &Operation) -> u8 {
    match op {
        Operation::Add => 1,
        Operation::Subtract => 1,
        Operation::Multiply => 2,
        Operation::Divide => 2,
        Operation::Exponentiate => 2,
    }
}



#[derive(Clone)]
enum Expression<'a> {
    BinOp(&'a Expression<'a>, Operation, &'a Expression<'a>),
    Num(u32),
}

fn pratt_parse(
    current_precedence: u8,
    tokens: &RefCell<VecDeque<Token>>,
) -> Option<Expression<'static>> {
    match tokens.borrow_mut().pop_front()? {
        Token::Op(_) => None,
        Token::Num(num) => {
            let left = Cell::new(Expression::Num(num));
            while tokens.borrow().len() != 0 {
                match tokens.borrow_mut().get(0).unwrap() {
                    Token::Op(op) => {
                        let prec = precedence(op);
                        if prec <= current_precedence {
                            let unwrapped = left.into_inner();
                            return Some(unwrapped);
                        }
                        tokens.borrow_mut().pop_front();
                        let right = pratt_parse(prec, tokens)?;
                        let unwrapped = left.into_inner();
                        left.replace(
                            Expression::BinOp(&unwrapped, *op, &right.clone())
                        );
                    }
                    Token::Num(_) => {
                        let unwrapped = left.into_inner();
                        return Some(unwrapped);
                    }
                }
            }
            let unwrapped = left.into_inner();
            Some(unwrapped)
        }
    }
}

fn parse(tokens: Vec<Token>) -> Option<Expression<'static>> {
    if tokens.len() == 0 {
        return None;
    }

    let token_deque = RefCell::new(VecDeque::from(tokens));

    return pratt_parse(0, &token_deque);
}
